### ISSUE TITLE
Multiple Vehicle Contraction's lead to OOM when Deserialized

### ISSUE THREAD

**Issue by juliusfriedman** [itinero/routing#300]
Give anymore than 1 contraction (depending on the size of the original source) I am running into out of memory exception in `RouterDb.Deserialize(stream);`

I have not explored if `RouterDb.Deserialize(stream, RouterDbProfile.NoCache);` will work for me yet but I can attempt and let you know.

Adding a single contraction seems to work to resolve the issues I and other experience in https://github.com/itinero/routing/issues/256 so adding multiple contractions (1 for each profile) is desired to resolve this for all cases where the API can be called with different profiles, e.g. a contraction for Fastest and Shortest per Vehicle.

I am only working with 2 vehicles so there is a total of 4 contractions of a roughly 350mb osm.pfb file, which results in a 1.2 gb file and a 500 mb with only a single contraction. The single contraction version loads up file however the version with 2 or more results in OOM.

I suppose I will have to keep the stream open and use `RouterDb.Deserialize(stream, RouterDbProfile.NoCache);` and hopefully that will work, I just wanted to let you know

**Comment by xivk** [comment]
Thanks, the OOM is strange, it should load when you have the available memory, perhaps an issue with the routerdb code itself.

If you can send me the routerdb that fails to load I can check myself.

Also, make sure you run using and x64 profile.

**Comment by juliusfriedman** [comment]
I was using anycpu so I apolagize for that. I will try again with x64 to ensure it's not that.

Thank you!

**Comment by juliusfriedman** [comment]
I confirmed this is still occurring in x64, I have the 1.3 GB routerdb file if your interested, just need to how to share, drop box or ftp etc.

**Comment by xivk** [comment]
Whatever you can manage is fine...

**Comment by juliusfriedman** [comment]
https://www.dropbox.com/s/0xf3a8t97txq8rh/planet-latest-filtered.contracted.routerdb?dl=0

Created from this osm.pbf https://www.dropbox.com/s/uoml5w7nare3izm/planet-latest-filtered.osm.pbf?dl=0

**Comment by juliusfriedman** [comment]
I still have not tried with RouterDb.NoCache yet.

Also please see https://github.com/itinero/routing/issues/256 as with multiple Contractions I am getting some weird exceptions during Multi-threading.

**Comment by juliusfriedman** [comment]
Just wondering if you had a chance to look at this yet.

I can take a look soon but since my multi threading issues are solved in https://github.com/itinero/routing/pull/302 I am not sure if we need contractions..

Sometimes with contractions we see better routes other times not so so some more tests still need to be performed to determine how useful they are for us as the RouterDb seems fast enough with or without the contractions.

I may end up having separate routerDb instances, one contracted and one not and perform analyzation using both outputs somehow.

I will definitely take a look at this when I get more time though. 

Hopefully I will get time this weekend.

**Comment by juliusfriedman** [comment]
I think the nuget package's are being built for x86 or any cpu and is why I encountered this issue. When I build from the source (develop) I don't seem to see this issue. Going to run a few tests today to attempt to load the same DB I submitted to you on the same computer which I previously saw the error (only changing the nuget for the source build) and see if I can still replicate this, if not I will close.

thank you
